src/lib/Occurrence.hs

```diff
@@ -0,0 +1,623 @@
+
+-- === Use Analysis ===
+
+-- We want to know how a binding is used in a Dex program, so as to
+-- know whether inlining it is guaranteed to conserve work and code
+-- size.
+
+-- Conservation of code size is easy.  Coservation of work, however,
+-- is more subtle.  Inlining the computation of x into
+--   for i. x
+-- does not conserve work (unless the index set has size 0 or 1), but
+-- inlining x into
+--   for i. x.i
+-- does conserve work -- provided we have access to the loop body
+-- that constructs it so we can beta-reduce against the indexing
+-- operation.
+
+-- The main difficulty is that Dex allows indexing by expressions
+-- other than referencing a loop binder, so to solve subtle cases we
+-- need to translate iteration-space expressions into data-space.
+-- For example, inlining x into
+--   for i. x.(Left i)
+--   for j. x.(Right j)
+-- conserves work, because the two injections have disjoint ranges.
+
+-- We proceed with a four-phase translation.
+--
+-- - Phase 1 translates Dex source into the Access structure.  An
+--   Access mirrors the control pattern of the Dex source, abstracting
+--   away most Dex syntax but still referring to iteration space.
+--   Access retains nearly complete usage information.
+--
+-- - Phase 2 transposes Access into Use by turning iteration space
+--   into data space.  (Use is like Access, except organized in data
+--   space.)  Use retains about the same amount of usage information
+--   as Access.
+--
+-- - Phase 3 collapses Use into CollapsedUse.  This process normalizes
+--   the subsets of the data space that are actually accessed, such as
+--   realizing that the subsets of all `Left` values and all `Right`
+--   values cannot intersect.  This is where most of the
+--   overapproximations are; CollapsedUse forgets much of the
+--   detailed usage information Use retained.
+--
+-- - Phase 4 actually extracts the inlining-relevant information from
+--   CollapsedUse.
+
+-- === Algebra ===
+
+-- The algebraic structure we want in order to represent uses is
+-- max-plus: max corresponds to `if` or `case` in the Dex program, and
+-- plus corresponds to multiple sequential statements.
+--
+-- This is almost a semiring, in that
+-- - max and plus are both monoids
+-- - max (and plus) are commutative
+-- - plus distributes over max
+-- - however, in our case, all actual uses that will occur are
+--   "positive", which means we can use `zero` (the identity for plus)
+--   as the identity for max as well, except that it then doesn't obey
+--   the annihilation law.
+
+class MaxPlus a where
+  zero :: a
+  max :: a -> a -> a
+  plus :: a -> a -> a
+
+-- === Access ===
+
+-- We represent an upper bound on the pattern by which (the object
+-- bound to) a name is accessed.  The goal is to determine whether
+-- inlining the name could duplicate work, so this is an upper bound
+-- (as opposed to an estimate or a lower bound).
+
+-- Each leaf or subtree of the object may be accessed a known or
```

Comment:
What is a "leaf or subtree of the object"?

Reply:
Until we generalize to tuple-typed accessees, this just means element or subarray.  Comment adjusted.